libpod/image/pull.go

```diff
@@ -114,64 +114,113 @@ func (ir *Runtime) getSinglePullRefPairGoal(srcRef types.ImageReference, destNam
+		destNames, err := reader.ManifestTagsForReference(ref)
 		if err != nil {
 			return nil, err
 		}
-		defer tarSource.Close()
-		manifest, err := tarSource.LoadTarManifest()
 
-		if err != nil {
-			return nil, errors.Wrapf(err, "error retrieving manifest.json")
-		}
-		// to pull the first image stored in the tar file
-		if len(manifest) == 0 {
-			// use the hex of the digest if no manifest is found
-			reference, err := getImageDigest(ctx, srcRef, sc)
+		if len(destNames) == 0 {
+			destName, err := getImageDigest(ctx, ref, sc)
 			if err != nil {
 				return nil, err
 			}
-			return ir.getSinglePullRefPairGoal(srcRef, reference)
+			destNames = append(destNames, destName)
+		} else {
+			for i := range destNames {
+				ref, err := NormalizedTag(destNames[i])
+				if err != nil {
+					return nil, err
+				}
+				destNames[i] = ref.String()
+			}
 		}
 
-		if len(manifest[0].RepoTags) == 0 {
-			// If the input image has no repotags, we need to feed it a dest anyways
-			digest, err := getImageDigest(ctx, srcRef, sc)
+		for _, destName := range destNames {
+			destRef, err := is.Transport.ParseStoreReference(ir.store, destName)
 			if err != nil {
-				return nil, err
+				return nil, errors.Wrapf(err, "error parsing dest reference name %#v", destName)
 			}
-			return ir.getSinglePullRefPairGoal(srcRef, digest)
+			if srcRef != nil {
+				ref = srcRef
+			}
+			pair := pullRefPair{
+				image:  destName,
+				srcRef: ref,
+				dstRef: destRef,
+			}
+			refPairs = append(refPairs, pair)
 		}
+	}
 
-		// Need to load in all the repo tags from the manifest
-		res := []pullRefPair{}
-		for _, dst := range manifest[0].RepoTags {
-			//check if image exists and gives a warning of untagging
-			localImage, err := ir.NewFromLocal(dst)
-			imageID := strings.TrimSuffix(manifest[0].Config, ".json")
-			if err == nil && imageID != localImage.ID() {
-				logrus.Errorf("the image %s already exists, renaming the old one with ID %s to empty string", dst, localImage.ID())
-			}
+	return refPairs, nil
+}
 
-			pullInfo, err := ir.getPullRefPair(srcRef, dst)
-			if err != nil {
-				return nil, err
+// getPullRefPairFromDockerReference returns a slice of pullRefPairs for the specified docker reference.
+func (ir *Runtime) getPullRefPairsFromDockerReference(ctx context.Context, ref types.ImageReference, sc *types.SystemContext) ([]pullRefPair, error) {
+	// Let's first assume it's a file-only reference.
+	reader, err := archive.NewReader(sc, ref.StringWithinTransport())
+	if err == nil {
+		defer func() {
+			if err := reader.Close(); err != nil {
+				logrus.Errorf(err.Error())
 			}
-			res = append(res, pullInfo)
+		}()
+		return ir.getPullRefPairsFromDockerArchiveReader(ctx, reader, ref, sc)
+	}
+
+	var destName string
+	// Determine the destination. If it's a tagged item, we
+	// can use it directly, otherwise we need to determine
+	// the image digest.
+	if dkrRef := ref.DockerReference(); dkrRef != nil {
+		destName = ref.DockerReference().String()
```

Comment:
Doesn’t this change behavior to always use the `docker.io/library` normalization? I’d expect `reader.ManifestTagsForReference` to be invoked here.

Reply:
Fixed